### PR TITLE
Variable length path

### PR DESCRIPTION
**Pull request by swilly22**

### CODE REVIEW

File: src/parser/grammar.y [hunk 1]
```diff
@@ -216,22 +217,54 @@ link(A) ::= LEFT_ARROW edge(B) DASH . {
...
 }
 
 // Edge with alias and label [alias:label]
-edge(A) ::= LEFT_BRACKET UQSTRING(B) COLON UQSTRING(C) properties(D) RIGHT_BRACKET . { 
-	A = New_AST_LinkEntity(B.strval, C.strval, D, N_DIR_UNKNOWN);
+edge(A) ::= LEFT_BRACKET UQSTRING(B) COLON UQSTRING(C) edgeLength(D) properties(E) RIGHT_BRACKET . { 
+	A = New_AST_LinkEntity(B.strval, C.strval, E, N_DIR_UNKNOWN, D);
+}
+
+// Edge with length [alias:*1..3]
```
Comment (jeffreylovitz) [inline src/parser/grammar.y]: Cypher doesn't support edges that have aliases but no labels in general (a change we implemented about a month ago, I think), which remains true with length specifiers:
```
MATCH (a)-[alias:*]->(b) RETURN b
> Invalid input '*': expected whitespace or a rel type name
```
We should, however, support edges with labels and length specifications, like:
`MATCH (a)-[:KNOWS*..2]->(b) RETURN b`

The Cypher Query Language Reference (version 9, pages 11-12) doesn't explicitly mention edges with labels, aliases, and length specifications, but they are supported in Neo.
`MATCH (a)-[alias:KNOWS*]->(b) RETURN b`

File: src/parser/grammar.y [hunk 2]
```diff
@@ -216,22 +217,54 @@ link(A) ::= LEFT_ARROW edge(B) DASH . {
...
+edge(A) ::= LEFT_BRACKET UQSTRING(B) COLON edgeLength(C) properties(D) RIGHT_BRACKET . {
```
Comment (jeffreylovitz) [inline src/parser/grammar.y]: `edgeLength` should not be preceded by a colon.

File: src/parser/grammar.y [hunk 3]
```diff
@@ -216,22 +217,54 @@ link(A) ::= LEFT_ARROW edge(B) DASH . {
...
+	A = New_AST_LinkEntity(B.strval, C.strval, E, N_DIR_UNKNOWN, D);
+}
+
+// Edge with length [alias:*1..3]
+edge(A) ::= LEFT_BRACKET UQSTRING(B) COLON edgeLength(C) properties(D) RIGHT_BRACKET . { 
+	A = New_AST_LinkEntity(B.strval, NULL, D, N_DIR_UNKNOWN, C);
+}
+
+// Edge with length [:*1..3]
+edge(A) ::= LEFT_BRACKET COLON edgeLength(B) properties(C) RIGHT_BRACKET . {
```
Comment (jeffreylovitz) [inline src/parser/grammar.y]: `edgeLength` should not be preceded by a colon.

File: src/parser/grammar.y [hunk 4]
```diff
@@ -216,22 +217,54 @@ link(A) ::= LEFT_ARROW edge(B) DASH . {
...
+
+%type edgeLength {AST_LinkLength*}
+
+// Edge length not specified.
+edgeLength(A) ::= . {
+	A = NULL;
+}
+
+// *minHops..maxHops
+edgeLength(A) ::= MUL INTEGER(B) DOT DOT INTEGER(C). {
```
Comment (jeffreylovitz) [inline src/parser/grammar.y]: The `DOT DOT` logic isn't working for me:
```
127.0.0.1:6379> GRAPH.QUERY social "MATCH (a)-[:*1..2]->(b) RETURN a, b"
(error) Syntax error at offset 18 near '.2'
```
As a guess, it may be trying to parse a double instead of an int?

File: src/parser/grammar.y [hunk 5]
```diff
@@ -216,22 +217,54 @@ link(A) ::= LEFT_ARROW edge(B) DASH . {
...
+	A = New_AST_LinkLength(B.intval, C.intval);
+}
+
```
Comment (jeffreylovitz) [inline src/parser/grammar.y]: It should also be legal to omit a bound, as in:
`(a)-[*..3]->(b)`
or
`(a)-[*3..]->(b)`

File: src/parser/ast_common.h [hunk 6]
```diff
@@ -35,13 +35,20 @@ typedef struct {
 	char *property;
 } AST_Variable;
 
+typedef struct {
+	unsigned int minHops;
+	float maxHops;
```
Comment (jeffreylovitz) [inline src/parser/ast_common.h]: I'm guessing we're using a float for `maxHops` to provide simpler support for `*` (infinite) length paths? That makes sense, and is probably fine, but I wouldn't mind defining UINT_MAX as an infinity value either.

The smallest integer value that can't be represented as a float is 2^24 + 1, which is not a path length we reasonably need to be concerned about, so it's not a big deal - your call!

File: src/execution_plan/ops/op_var_len_traverse.c [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+#include "./op_var_len_traverse.h"
+
+void _VarLenTraverseOpHop(VarLenTraverse *op) {
+    GrB_mxm(
+        op->frontier,           // Output
+        op->mask,               // Mask
+        NULL,                   // Accumulator
+        GxB_LOR_LAND_BOOL,      // Semiring
+        op->A,                  // First matrix
+        op->A,                  // Second matrix
```
Comment (jeffreylovitz) [inline src/execution_plan/ops/op_var_len_traverse.c]: I don't understand what this does! We use the same matrix for both inputs?

File: src/execution_plan/ops/op_var_len_traverse.c [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    return OP_OK;
+}
+
+OpResult VarLenTraverseReset(OpBase *ctx) {
+    VarLenTraverse *op = (VarLenTraverse*)ctx;
+    op->hops = 1;
+    GrB_free(&op->mask);
+    GrB_Matrix_new(&op->mask, GrB_BOOL, Graph_NodeCount(op->g), Graph_NodeCount(op->g));
+    GrB_Matrix_dup(&op->mask, op->A);
+    TuplesIter_clear(op->it);
```
Comment (jeffreylovitz) [inline src/execution_plan/ops/op_var_len_traverse.c]: What does this accomplish? The iterator will return DEPLETED on its next invocation, but it does not actually get invoked before the `TuplesIter_reuse` call on 107.

File: src/parser/grammar.y (hunk 1 above)
Comment (swilly22) [reply src/parser/grammar.y]: You're correct, although I don't see a reason why we shouldn't support this
I think its quite powerful to ask (A)-[:*1..20]->(B)

File: src/parser/grammar.y (hunk 4 above)
Comment (swilly22) [reply src/parser/grammar.y]: introducing DOTDOT

File: src/parser/grammar.y (hunk 1 above)
Comment (jeffreylovitz) [reply src/parser/grammar.y]: @swilly22 As a minor clarification, `(A)-[*1..20]->(B)` (returning all matching paths without providing an alias to refer to those paths in other parts of the query) is totally canonical syntax, and we should support it!

Cypher does not explicitly support `(A)-[alias:*1..20]->(B)`, and - conjecturing - I suspect that is because it's proven computationally challenging to merge the properties of all matching paths into a single edge alias that can be filtered, returned, and so forth. I can't think of any syntactical reason for this constraint other than that it taxes the system, especially given that Cypher provides support for multiple explicit edge labels with syntax like `(a)-[:KNOWS|HAS_MET)->(b)`.

I have no objection to going beyond the language specification, but providing support for edge aliases that don't correspond to specific labels will introduce a fair few cases we'll need to add tooling for. As an example, if one wants to filter on a property found along a path, we'll need to find and track the properties of _all_ paths matching the query, as opposed to the masking approach suggested in this PR, which at each hop only considers edges that can modify the paths being built. I'm specifically referring to the `_VarLenTraverseOpHop` logic that builds and employs a mask to ignore traversals that, on a boolean level, have been made redundant, but I expect that is not the only edge case we'd need to compensate for.

This is just food for thought - I'm on board with providing more functionality whenever someone might find it useful! Let me know what you think.

File: tests/unit/test_all_paths.cpp [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+ */
+
+#include "../../deps/googletest/include/gtest/gtest.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "../../src/algorithms/algorithms.h"
+#include "../../src/util/arr.h"
```
Comment (jeffreylovitz) [inline tests/unit/test_all_paths.cpp]: I think we need to move the function definitions in `arr.h` into a separate .c file before these tests will compile correctly for me/Circle.

File: src/parser/grammar.y [hunk 10]
```diff
@@ -216,22 +217,54 @@ link(A) ::= LEFT_ARROW edge(B) DASH . {
...
+	A = New_AST_LinkEntity(NULL, B.strval, D, N_DIR_UNKNOWN, C);
 }
 
 // Edge with alias and label [alias:label]
-edge(A) ::= LEFT_BRACKET UQSTRING(B) COLON UQSTRING(C) properties(D) RIGHT_BRACKET . { 
-	A = New_AST_LinkEntity(B.strval, C.strval, D, N_DIR_UNKNOWN);
+edge(A) ::= LEFT_BRACKET UQSTRING(B) COLON UQSTRING(C) edgeLength(D) properties(E) RIGHT_BRACKET . { 
+	A = New_AST_LinkEntity(B.strval, C.strval, E, N_DIR_UNKNOWN, D);
+}
+
```
Comment (jeffreylovitz) [inline src/parser/grammar.y]: These rules look good! We still need to add one for edges that just have length specifiers, and no alias or label.

File: src/parser/ast_common.c [hunk 11]
```diff
@@ -37,6 +38,13 @@ AST_LinkEntity* New_AST_LinkEntity(char *alias, char *label, Vector *properties,
 	return le;
 }
 
+AST_LinkLength* New_AST_LinkLength(unsigned int minHops, unsigned int maxHops) {
```
Comment (jeffreylovitz) [inline src/parser/ast_common.c]: It would be good here or elsewhere to verify that minHops <= maxHops

File: src/graph/graph.h [hunk 12]
```diff
@@ -101,11 +101,19 @@ void Graph_GetEdgesConnectingNodes (
     Vector *edges
 );
 
-// Get both incoming and outgoing node edges.
+// Get node edges.
```
Comment (jeffreylovitz) [inline src/graph/graph.h]: Comment belongs above `Graph_GetNodeEdges`

File: src/graph/graph.c [hunk 13]
```diff
@@ -617,9 +623,15 @@ GrB_Matrix Graph_GetLabel(const Graph *g, int label_idx) {
 }
 
 GrB_Matrix Graph_GetRelation(const Graph *g, int relation_idx) {
```
Comment (jeffreylovitz) [inline src/graph/graph.c]: `size_t relation_idx`

File: src/execution_plan/ops/op_cond_var_len_traverse.c [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+
+    OpResult res;
+    GrB_Index src_id;
+    GrB_Index dest_id;
+
+    /* Not initialized. */
+    while(array_len(op->paths) == 0) {
+        res = child->consume(child, graph);
+        if(res != OP_OK) return res;
+        printf("op->maxHops: %d\n", op->maxHops);
```
Comment (jeffreylovitz) [inline src/execution_plan/ops/op_cond_var_len_traverse.c]: This is of course just a debug line, but the print specifier should be `%u`

File: src/algorithms/path.c [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    p = array_append(p, e);
+    return p;
+}
+
+Edge *Path_pop(Path p) {
+    size_t pathLen = array_len(p);
+    if(pathLen == 0) return NULL;
+    
+    Edge *e = p[pathLen-1];
+    array_pop(p);
```
Comment (jeffreylovitz) [inline src/algorithms/path.c]: These two lines are equivalent to:
`Edge *e = array_pop(p);`
Or can even `return array_pop(p)`

File: src/algorithms/path.c [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+#include "../util/arr.h"
+
+Path Path_new(size_t len) {
+    Path p = array_new(Edge*, len);
+    return p;
+}
+
+Path Path_append(Path p, Edge *e) {
+    p = array_append(p, e);
+    return p;
```
Comment (jeffreylovitz) [inline src/algorithms/path.c]: Can simply `return array_append`

File: src/algorithms/path.c [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+* This file is available under the Apache License, Version 2.0,
+* modified with the Commons Clause restriction.
+*/
+
+#include "./path.h"
+#include "../util/arr.h"
+
+Path Path_new(size_t len) {
+    Path p = array_new(Edge*, len);
+    return p;
```
Comment (jeffreylovitz) [inline src/algorithms/path.c]: Can simply `return array_new`

File: src/algorithms/all_paths.c [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+            GrB_Matrix_setElement_BOOL(visited, true, neighborID, frontierID);
+
+            *path = Path_append(*path, e);
+
+            _AllPaths(g, Edge_GetDestNode(e), relationID, relation, minHops, maxHops, hop+1, visited, path, paths);
+
+            Path_pop(*path);
+
+            // Mark edge as unvisited.
+            GrB_Matrix_setElement_BOOL(visited, false, neighborID, frontierID);
```
Comment (jeffreylovitz) [inline src/algorithms/all_paths.c]: This probably necessitates a zero vector or `GxB_NONZERO` reduction later on, right?

Comment (jeffreylovitz) [reply src/algorithms/all_paths.c]: Actually, never mind - I suppose it does not so long as we never try to iterate over visited with a TuplesIter or similar.

File: tests/unit/test_all_paths.cpp (hunk 9 above)
Comment (swilly22) [reply tests/unit/test_all_paths.cpp]: Already solved, simple casting.

File: src/algorithms/all_paths.c (hunk 18 above)
Comment (swilly22) [reply src/algorithms/all_paths.c]: visited, won't be iterated, it only serves as a marker for checking if an edge been traversed.

File: src/graph/graph.c [hunk 19]
```diff
@@ -392,43 +392,49 @@ void Graph_GetEdgesConnectingNodes(const Graph *g, NodeID src, NodeID dest, int
 
 /* Retrieves all either incoming or outgoing edges 
  * to/from given node N, depending on given direction. */
-void Graph_GetNodeEdges(const Graph *g, const Node *n, Vector *edges) {
+void Graph_GetNodeEdges(const Graph *g, const Node *n, Vector *edges, GRAPH_EDGE_DIR dir, int edgeType) {
     assert(g && n && edges);
     NodeID srcNodeID;
     NodeID destNodeID;
     TuplesIter *tupleIter;
```
Comment (jeffreylovitz) [inline src/graph/graph.c]: `TuplesIter *tupleIter = NULL` will suppress a compiler warning for me.

File: src/algorithms/path.c [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+    size_t pathLen = array_len(p);
+    if(pathLen == 0) return NULL;
+    return array_pop(p);
+}
+
+size_t Path_len(const Path p) {
+    return array_len(p);
+}
+
+Path Path_clone(const Path p) {
```
Comment (jeffreylovitz) [inline src/algorithms/path.c]: It seems to me like we could just perform a memcpy on the array object to accomplish this in one instruction (arr.h provides an `array_sizeof` macro already). That's not particularly important, though, we can just keep it in mind for the future.

File: src/algorithms/path.c [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+    for (int i = 0; i < pathLen; i++) {
+        clone = Path_append(clone, p[i]);
+    }
+
+    return clone;
+}
+
+void Path_print(Path p) {
+    assert(p);
+    Edge *e;
```
Comment (jeffreylovitz) [inline src/algorithms/path.c]: `Edge *e = NULL` will suppress a compiler warning for me.
